jetty-util/src/main/java/org/eclipse/jetty/util/Index.java

```diff
@@ -0,0 +1,148 @@
+//
+
+package org.eclipse.jetty.util;
+
+import java.nio.ByteBuffer;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+
+public interface Index<V> extends OpaqueIndex
+{
+    @Override
+    default boolean contains(String key)
+    {
+        return get(key) != null;
+    }
+
+    @Override
+    default boolean contains(String key, int offset, int length)
+    {
+        return get(key, offset, length) != null;
+    }
+
+    /**
+     * Get an exact match from a String key
+     *
+     * @param s The key
+     * @return the value for the string key
+     */
+    V get(String s);
+
+    /**
+     * Get an exact match from a String key
+     *
+     * @param s The key
+     * @param offset The offset within the string of the key
+     * @param len the length of the key
+     * @return the value for the string / offset / length
+     */
+    V get(String s, int offset, int len);
+
+    /**
+     * Get the best match from key in a String.
+     *
+     * @param s The string
+     * @param offset The offset within the string of the key
+     * @param len the length of the key
+     * @return The value or null if not found
+     */
+    V getBest(String s, int offset, int len);
+
+    /**
+     * Get the best match from key in a byte buffer.
+     * The key is assumed to by ISO_8859_1 characters.
+     *
+     * @param b The buffer
+     * @param offset The offset within the buffer of the key
+     * @param len the length of the key
+     * @return The value or null if not found
+     */
+    V getBest(ByteBuffer b, int offset, int len);
+
+    class Builder<V>
+    {
+        private Map<String, V> contents = new LinkedHashMap<>();
+        private boolean caseSensitive;
+
+        public Builder<V> caseSensitive(boolean caseSensitive)
+        {
+            this.caseSensitive = caseSensitive;
+            return this;
+        }
+
+        public <T> Builder<V> withAllDerivate(T[] values, Function<T, Map<String, V>> function)
+        {
+            for (T value : values)
+            {
+                Map<String, V> pairs = function.apply(value);
+                contents.putAll(pairs);
+            }
+            return this;
+        }
+
+        public Builder<V> withAll(V[] values, Function<V, String> function)
+        {
+            for (V value : values)
+            {
+                String key = function.apply(value);
+                contents.put(key, value);
+            }
+            return this;
+        }
+
+        public Builder<V> with(String key, V value)
+        {
+            contents.put(key, value);
+            return this;
+        }
+
+        public Index<V> build()
+        {
+            int capacity = 16;
+            Trie<V> result = createTrie(capacity);
+            for (Map.Entry<String, V> entry : contents.entrySet())
+            {
+                while (!result.put(entry.getKey(), entry.getValue()))
```

Comment:
I think I've found a "simple" algorithm to work out the required capacity from a set of keys:
```java
    public static int requiredCapacity(Set<String> keys)
    {
        List<String> list = new ArrayList<>(keys);
        Collections.sort(list);

        return requiredCapacity(list, 0, list.size(), 0);
    }

    private static int requiredCapacity(List<String> keys, int offset, int length, int index)
    {
        if (length == 0)
            return 0;

        int required = 0;

        Character c = null;
        for (int i = 0; i < length; i++)
        {
            String k = keys.get(offset + i);

            if (c == null)
            {
                if (k.length() >= index + 1)
                {
                    required++;
                    c = k.charAt(index);;
                }
                else
                {
                    offset++;
                    length--;
                    i--;
                }
                continue;
            }

            if (k.length() >= index + 1)
            {
                char n = k.charAt(index);
                if (c != n)
                {
                    required += requiredCapacity(keys, offset, i, index + 1) + 1;
                    offset = i;
                    length -= i;
                    c = n;
                    i = 1;
                }
                continue;
            }

            c = null;
            offset = i;
            length -= i;
            i = 0;
        }

        required += requiredCapacity(keys, offset, length, index + 1);
        return required;
    }
```

Which passes the following tests:

```java
    @Test
    public void testArrayTrieRequestCapacity()
    {
        assertThat(ArrayTrie.requiredCapacity(Set.of("ABC")), is(3));
        assertThat(ArrayTrie.requiredCapacity(Set.of("ABC", "XYZ")), is(6));
        assertThat(ArrayTrie.requiredCapacity(Set.of("A00", "A11")), is(5));
        assertThat(ArrayTrie.requiredCapacity(Set.of("A00", "A01", "A10", "A11")), is(7));
        assertThat(ArrayTrie.requiredCapacity(Set.of("ABC", "ABCDEF")), is(6));
        assertThat(ArrayTrie.requiredCapacity(Set.of("A", "AB")), is(2));
        assertThat(ArrayTrie.requiredCapacity(Set.of("AB", "A")), is(2));
        assertThat(ArrayTrie.requiredCapacity(Set.of("ABC", "ABCDEF")), is(6));
        assertThat(ArrayTrie.requiredCapacity(Set.of("ABCDEF", "ABC")), is(6));
        assertThat(ArrayTrie.requiredCapacity(Set.of("ABC", "ABCDEF", "ABX")), is(7));
        assertThat(ArrayTrie.requiredCapacity(Set.of("ABCDEF", "ABC", "ABX")), is(7));
    }
```

Reply:
Thanks! I'll have a look at it as it'll be needed to build immutable `Index`es.